cpu/efm32/uart: fix handling of RX when no RX callback is configured

benemorius opened the pull request:
### Contribution description
This fixes a crash that happens on efm32 when the UART receives a character while no UART RX callback is configured.

The first commit is adequate to avoid the crash but the next commits do some refactoring to make the rest of the driver consistent with the use case where no RX callback is configured.

### Testing procedure
1) Flash any example with a shell and confirm that UART still works as intended.
1) Flash the `hello-world` example (it has no shell) and confirm that typing UART characters causes no crash
1) Without this PR, the previous step causes a crash due to calling `rx_cb()` while `rx_cb == NULL`

basilfx commented on cpu/efm32/periph/uart.c:
```diff
@@ -131,15 +129,16 @@ int uart_init(uart_t dev, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
 
         /* enable receive interrupt */
         LEUART_IntEnable(leuart, LEUART_IEN_RXDATAV);
```
I guess this also needs to be wrapped inside an if-statement

benemorius replied:
What I did is leave the interrupt always enabled in the peripheral and put only an if-block around enabling the interrupt in the NVIC. I don't think there would be a functional change but I'm mostly indifferent about it and I'll add if-blocks if you disagree.

basilfx commented on cpu/efm32/periph/uart.c:
```diff
@@ -131,15 +129,16 @@ int uart_init(uart_t dev, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     /* enable the interrupt */
-    NVIC_ClearPendingIRQ(uart_config[dev].irq);
-    NVIC_EnableIRQ(uart_config[dev].irq);
+    if (isr_ctx[dev].rx_cb) {
```
Use `rx_cb` instead.

basilfx commented on cpu/efm32/periph/uart.c:
```diff
@@ -164,10 +163,62 @@ void uart_write(uart_t dev, const uint8_t *data, size_t len)
+        USART_TypeDef *usart = uart_config[dev].dev;
+
+        /* enable tx */
+        usart->CMD = USART_CMD_TXEN;
```
Please use `USART_Enable` (and `LEUART_Enable`). There is [more happening](http://devtools.silabs.com/dl/documentation/doxygen/5.7/efr32mg12/html/em__usart_8c_source.html#l00663) in these methods to have compatibility with other EFM32 families.

benemorius replied:
The emlib functions can't be used to enable TX and RX independently though right?

benemorius replied:
Nevermind. They can of course.

basilfx commented:
@benemorius thanks for the fixes. I have a few comments about the `uart_poweron`/`uart_poweroff` methods.

I was able to reproduce this issue with latest `master` on the STK3600.

benemorius commented:
Thanks I addressed the comments except for one pending a response.

Also sorry I added one more commit to address the case where `uart_write()` is followed immediately by `uart_poweroff()` by blocking until TX is complete. This is consistent with the stm32 driver for example.

basilfx commented:
@benemorius friendly reminder :-)

benemorius commented:
Very sorry! I'll be back this week.

basilfx commented:
Gave this another go on the STK3600, SLSTK3402a and my own custom board, and all is good. I'm happy with the changes, so go!
